proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1100 @@
+``useC``).
+
+On the other hand, the following program is rejected:
+
+::
+
+   neglecting :: C %1 => Int -> Int
+   neglecting n = n + n
+
+as ``C`` is never consumed.
+
+Furthermore, as with linear arguments, we must *guarantee* that ``C``
+will be consumed, regardless of the conditional branching. Thus, the
+following would also be rejected:
+
+::
+
+   dithering :: C %1 => Bool -> Int
+   dithering x = if x then useC else 10
+as ``useC`` is only consumed when ``x == True``.
+
+These examples show programs rejected by failing to guarantee that
+``C`` will be used. However, as we are in a linear context, we must
+also ensure that it isn't *overused*:
+
+::
+
+   overusing :: C %1 => (Int, Int)
+   overusing = (useC, useC)
+
+Since ``overusing`` consumes a linear ``C`` constraint twice, it is
+rejected for violating linearity. We may amend the type scheme of
+``overusing`` in the following way, however:
+
+::
+
+   stammering :: (C, C) %1 => (Int, Int)
+   stammering = (useC, useC)
+
+By providing an additional linear constraint, the resource usage
+guarantees are satisfied and the program type checks.
+
+The theoretical basisof this proposal, as well as further examples,
+can be found in the `Linearly Qualified Types paper <paper_>`_ by
+Spiwack et al.
+
+Motivation
+----------
+
+The motivation of this proposal is to remove unnecessary boilerplate
+that is often introduced when programming with linear types.
+
+Threading
+^^^^^^^^^
+
+Consider the following program:
+
+::
+
+   read2AndDiscard :: MArray a %1 -> (Ur a, Ur a)
+   read2AndDiscard arr0 =
+     let (arr1, x) = read arr0 0
+         (arr2, y) = read arr1 1
+         () = free arr2
+     in (x, y)
+
+This is a function which takes an array as a linear argument, reads
+the first two elements, and then de-allocates it before returning the
+two elements. The fact that the array is a linear resource provides
+some nice guarantees which ensure memory safety. For example, there is
+no way we can go on to erroneously read from the array after
+de-allocating it. Instead, when ``arr0`` is used as an argument to
+``read``, ``read`` consumes it and returns the array under a new name
+``arr1``. This process then repeats before finally ``free``
+de-allocates the array.
+
+While this code ensures that we use the ``MArray`` in a memory-safe
+way, this re-naming process introduces boilerplate and becomes
+cumbersome to both write and read. Compare to the doing the same with
+the ``ST`` monad instead:
+
+::
+
+   read2AndDiscardST :: MArray s a -> ST s (a, a)
+   read2andDiscardST arr = do
+     x <- read arr 0
+     y <- read arr 1
+     return (x, y)
+
+There is much less noise. To be honest, this is all a little silly:
+the name management is completely mechanical; it should be handled
+automatically by the compiler. This is what this proposal lets us do:
+get pretty close to the latter style, without having to introduce a
+monad.
```

Comment:
I'm taking this kind of for granted, because linear constraints or not, the whole point of this array example is to avoid monads. The simple fact that they are considered appealing by anyone (in fact it's the first example that people think of when they think of linear types, and gets them excited (though it's always been my position that it's not the most compelling example)) is evidence that mutation without monad is attractive.

That being said, the motivation section does link to [this blog post of mine](https://www.tweag.io/blog/2023-01-26-linear-constraints-freeze/) that gives you an example (recursive freezing) that seems absolute madness without linear constraints (even in the ST monad!).

Reply:
One more cool thing that is possible with linear constraints and isn't possible with vanilla IO/ST monad is to control that we don't use mutable buffer after deallocation/handle after closing or something similar.

A couple of months ago, I caught a bug with vanilla file handle API when I attempt to write into handle opened into ReadWrite mode after `hGetContents` function and got exception about closed handle. 

```haskell
withFile "file-name" ReadWriteMode \h -> do
  content <- Bytestring.Strict.hGetContents h
  hPutStr h "write into handle"
  pure content
```
It is turned out that `hGetContents` closes handle at the end.

With linearity-based API `hGetContents` just would not return handle Read/Write constraints, so bad code just would not compile.